src/Core/Renderer.cpp

```diff
@@ -97,45 +97,81 @@ void Raytracer::Renderer::renderFinalScene()
+
+    double supersampledWidth = imageWidth * supersamplingFactor;
+    double supersampledHeight = imageHeight * supersamplingFactor;
+
+    auto imageRatio = supersampledWidth / supersampledHeight;
 
     // Field of view in degrees
     double fov = _camera.getFov();
     auto focalLength = 1.0;
     auto viewHeight = 2.0 * std::tan((fov * M_PI / 180.0) / 2.0) * focalLength;
     auto viewWidth = viewHeight * imageRatio;
 
+    std::vector<std::vector<Color>> supersampledImage(supersampledHeight, std::vector<Color>(supersampledWidth));
+
     // Calculate the vectors across the horizontal and vertical viewport edges.
     auto viewU = Math::Vector3D(viewWidth, 0, 0);
     auto viewV = Math::Vector3D(0, -viewHeight, 0);
 
     // Calculate the horizontal and vertical delta vectors from pixel to pixel.
-    auto pixelSizeU = viewU / imageWidth;
-    auto pixelSizeV = viewV / imageHeight;
+    auto pixelSizeU = viewU / supersampledWidth;
+    auto pixelSizeV = viewV / supersampledHeight;
 
     // Calculate the location of the upper-left pixel.
     auto viewUpper_left = _camera.getOrigin() - Math::Vector3D(0, 0, focalLength) - viewU / 2 - viewV / 2;
 
+    for (int y = supersampledHeight - 1; y >= 0; y--) {
+        for (int x = 0; x < supersampledWidth; x++) {
+            auto pixelCenter = viewUpper_left + (pixelSizeU * (x + 0.5)) + (pixelSizeV * (y + 0.5));
+            auto rayDirection = _camera.getOrigin() - pixelCenter;
+
+            rayDirection.rotateX(this->_camera.getRotation().x());
+            rayDirection.rotateY(this->_camera.getRotation().y());
+            rayDirection.rotateZ(this->_camera.getRotation().z());
+
+            Raytracer::Ray r(_camera.getOrigin(), rayDirection);
+            Color hitColor = _primitives.getColorPoint(r, _lights);
+
+            supersampledImage[y][x] = hitColor;
+        }
+    }
+
+    std::vector<std::vector<Color>> finalImage(imageHeight, std::vector<Color>(imageWidth));
+
+    for (int y = 0; y < imageHeight; y++) {
+        for (int x = 0; x < imageWidth; x++) {
+            Color averageColor(0, 0, 0);
+            int pixelCount = 0;
+
+            for (int supersamplingY = 0; supersamplingY < supersamplingFactor; supersamplingY++) {
+                for (int supersamplingX = 0; supersamplingX < supersamplingFactor; supersamplingX++) {
+                    int sourceY = y * supersamplingFactor + supersamplingY;
+                    int sourceX = x * supersamplingFactor + supersamplingX;
+                    averageColor += supersampledImage[sourceY][sourceX];
+                    pixelCount++;
+                }
+            }
+
+            averageColor /= pixelCount;
+            finalImage[y][x] = averageColor;
+        }
+    }
+
+
     if (stream.is_open()) {
 
         stream << "P3\n" << imageWidth << ' ' << imageHeight << "\n255\n";
 
-        for (int y = imageHeight; y >= 0; y--) {
+        for (int y = 0; y < imageHeight; y++) {
```

Comment:
Be careful, you inverse the image: 
```suggestion
        for (int y = imageHeight; y >= 0; y--) {
```